docs/api.md

```diff
@@ -6,6 +6,171 @@ challenges presented by JavaScript and web browsers.
 Please direct clarifications or observations of inaccuracy to [our issue
 tracker](https://github.com/iodide-project/iodide/issues/new).
 
+
+
+## `iodide.file`
+
+The `iodide.file` API provides convenience functions around working with files uploaded to the Iodide server in your notebook.
+
+
+
+### `iodide.file.save(data, fileName[, saveOptions])`
+
+Returns a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises), which, when resolved, will signal that `data` was uploaded to the server under the filename `fileName`.
+
+`data` (required) is any object or variable in the eval name space. It can be anything, really - an array of data, or a csv, or a binary blob.
+
+`filename` (required) is a string that represents the file name.
+
+The optional argument `saveOptions` has the following keys:
+
+- `overwrite` (optional, default `false`): if `true`, will overwrite whatever is at `fileName` with `data`. If `false`, will throw an error and halt the subsequent enqueued evaluation (similar to a syntax error).
+- `serializer` (optional): a function that transforms `data` to its final form.
+
+
+##### `iodide.file.save` patterns
+
+```javascript
+
+// upload a csv to the server. Assume the object csvData is an array of objects
+// representing a columnar dataset. This overwrites whatever is in `cached-data.csv`
+// because overwrite is set to true.
+
+const csvData = [{x1: 10, x2: 'test1'}, {x1: 20, x2: 'test2'}, ...]
+
+iodide.file.save(csvData, 'cached-data.csv' {
+  overwrite: true,
+  serializer: d3.csvFormat
+})
+
+// iodide.file.save works with ArrayBuffers as well.
+iodide.file.save(new ArrayBuffer(...), 'model-output.bin')
+```
+
+
+
+### `iodide.file.load(fileName, fileType[, variableName])`
+
+Returns a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises) that, when resolved, loads into the notebook the save file `fileName`. For each notebook can find a list of uploaded files in the [upload modal](#link-required) or on your notebook's [Revisions Page](#needs-link). You can also access any uploaded file through the fetch chunk, [following this pattern](https://iodide-project.github.io/docs/workflows/#uploading-data-to-an-iodide-notebook). For most use cases using [fetch chunks](https://iodide-project.github.io/docs/jsmd/#fetch-chunks-fetch) is preferable and more straightforward. In more dynamic programmatic cases, however, `iodide.file.load` can provide more nuanced workflows.
+
+`fileName` is the name of the file uploaded to the Iodide server. 
+
+`fileType` is the file type to handle. These are the same as the file types in [fetch chunks](https://iodide-project.github.io/docs/jsmd/#fetch-chunks-fetch): `js` (load this file as javascript, eg a library), `css` (load this file as a stylesheet, applying the styles to the report), `json` (load this file as json and parse into a javascript object), `text` (load this file as text), `blob` (load this file as binary).
```

Comment:
I wonder if we should make this optional and have some sort of sensible default behaviour where not specified: maybe load the file as json if it's json, load the file as a blob otherwise?

Reply:
Yeah, I'm of two minds about this. We make it so you have to explicitly say what you want in the fetch chunk, and I think in general that's a good policy because it reduces magic a bit and lets a user know exactly what's going on at a pretty minimal cost. My hope would be to keep this API as similar to fetch chunks as possible. Blobs are not particularly gentle for people just trying to load a CSV, I think, so it may be better to just default to `text` if we go this direction. 

It could be that we need `iodide.file.load` to have other conveniences like `iodide.file.loadText` and `iodide.file.loadJSON` or whatever, which just wraps `iodide.file.load`, and leave this function as-is.

For now, I'll keep this as-is, and we can discuss this further during our Thursday meeting.